### PR TITLE
[DataGrid] Fix performance issue when sorting columns

### LINKED ISSUE #1330: [XGrid] Sorting completly freeze
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

Sorting a column with 100,000 rows completely freeze the UI.

## Expected Behavior 🤔

Works as before

## Steps to Reproduce 🕹

Steps:

1. Open https://material-ui.com/components/data-grid/
2. Click on any column to sort
3. CPU at 100% forever

## Context 🔦

v4.0.0-alpha.23

### PR DESCRIPTION
**Pull request by dtassone**
fix #1330

### CODE REVIEW

**Comment by DanailH** [comment]
The fix is working but it is still the thread freezes for a couple of seconds before the sorting finishes. Maybe we can add a loader to indicate that something is happening? But either way, the end-user should be able to perform other actions on the page while the sorting is happening, no?
Maybe this is a bugger problem that needs to be looked at.

**Review by oliviertassinari — COMMENTED** [review]
1. The performance regression is definitely fixed 👍 
1. It seems that we might have the opportunity to add a test case here. As far as I remember, it's the second time we break the performance of the sorting (or unless it was the checkbox selection only last time 🤔). Maybe we could render one grid with 1,000 rows, sort it and assert the time it took is x10 below a threshold that HEAD would break?
1. Please take the time to detail each pull request. I have dove into the issue to understand and explain what went wrong, to see if there is something we can learn, for the whole team, in the future.

The current issue is in the following call order:

- sorting trigger
- each row calls `api.getCellParams`
- which calls `getBaseCellParams`
- which calls `getCellElement`
- which calls `x.querySelector(y)` => really slow

Screenshot 2021-04-08 at 12 39 14

**Comment by dtassone** [comment]
perf is subject to the browser engine so it varies with browsers.

File: packages/grid/x-grid/src/tests/sorting.XGrid.test.tsx [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
+import {
+  createClientRenderStrictMode,
+  // @ts-expect-error need to migrate helpers to TypeScript
+  fireEvent,
+  // @ts-expect-error need to migrate helpers to TypeScript
+  screen,
+  // @ts-expect-error need to migrate helpers to TypeScript
+  waitFor,
+} from 'test/utils';
+import { useData } from '../../../../storybook/src/hooks/useData';
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/sorting.XGrid.test.tsx:16]:
```suggestion
import { useData } from 'packages/storybook/src/hooks/useData';
```

**Comment by oliviertassinari** [comment]
> perf is subject to the browser engine so it varies with browsers.

I have tried something, please review

File: packages/grid/x-grid/src/tests/sorting.XGrid.test.tsx [hunk 2]
```diff
@@ -131,4 +140,40 @@ describe('<XGrid /> - Sorting', () => {
...
+      const header = screen
+        .getByRole('columnheader', { name: 'Currency Pair' })
+        .querySelector('.MuiDataGrid-colCellTitleContainer');
+
+      const t0 = performance.now();
+      fireEvent.click(header);
+      await waitFor(() => expect(document.querySelector('.MuiDataGrid-sortIcon')).to.not.be.null);
+      const t1 = performance.now();
+      const time = Math.round(t1 - t0);
+      expect(time).to.be.lessThan(200);
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/sorting.XGrid.test.tsx:176]: It's not flying, could we increase the threshold?

Screenshot 2021-04-14 at 12 39 39

https://app.circleci.com/pipelines/github/mui-org/material-ui-x/4060/workflows/b303d403-165d-4802-9bf1-17d115e3c05f/jobs/18637

Also cc @DanailH that reported the issue as well

Comment (dtassone) [reply packages/grid/x-grid/src/tests/sorting.XGrid.test.tsx:176]: Yes we can increase the limit and add an acceptable buffer.

Comment (oliviertassinari) [reply packages/grid/x-grid/src/tests/sorting.XGrid.test.tsx:176]: Done https://github.com/mui-org/material-ui-x/blame/43f710c6b6e6d43b2e3ef07bd4627bfee06a1021/packages/grid/x-grid/src/tests/sorting.XGrid.test.tsx#L226

**Pull request merged**